### DISCUSSION TITLE
jdk需要怎么配置才可以编译java文件，配置好了jdk环境变量，运行时无法找到javac

### DISCUSSION THREAD

**Discussion by zengfanjin** [criyle/go-judge#99] [Q&A] [answered]
jdk需要怎么配置才可以编译java文件，配置好了jdk环境变量，运行时无法找到javac

**Comment by criyle** [comment]
什么 jdk 版本，怎么装的

**Reply by zengfanjin** [reply]
jdk17 和 jdk 8 都试过，tar包解压的

**Reply by zengfanjin** [reply]
直接运行的docker 命令安装的这个项目，在docker容器内部和服务器上都安装了jdk

**Reply by criyle** [reply, marked as answer]
放在哪个位置呢，默认情况下沙箱的隔离环境里面只有部分文件夹，例如 `/bin` `/usr` ，如果需要其他位置，需要修改 `mount.yaml`

**Comment by ppaet** [comment]
你好，问题解决了吗，我已经配置好java了，/run接口的请求体应该怎么写，java不是得先打包编译才能运行

**Reply by ppaet** [reply]
我解决了。
POST:/run
{
    "cmd": [{
        "args": ["/usr/lib/jvm/jdk1.8.0_401/bin/javac", "Main.java"],
        "env": ["PATH=/usr/bin:/bin"],
        "files": [{
            "content": ""
        }, {
            "name": "stdout",
            "max": 10240
        }, {
            "name": "stderr",
            "max": 10240
        }],
        "cpuLimit": 10000000000,
        "memoryLimit": 104857600,
        "procLimit": 50,
        "copyIn": {
            "Main.java": {
                "content": "public class Main{ \n    public static void main(String[] args) { \n        System.out.println(\"Hello world\");\n    }\n}"
            }
        },
        "copyOutCached": ["Main.class"]
    }]
}
响应：
[
    {
        "status": "Accepted",
        "exitStatus": 0,
        "time": 881790000,
        "memory": 60272640,
        "runTime": 534163061,
        "files": {
            "stderr": "",
            "stdout": ""
        },
        "fileIds": {
            "Main.class": "5UDHKNDA"
        }
    }
]
获取文件ID
POST:/run
{
    "cmd": [{
        "args": ["/usr/lib/jvm/jdk1.8.0_401/bin/java", "Main"],
        "env": ["PATH=/usr/bin:/bin"],
        "files": [{
            "content": ""
        }, {
            "name": "stdout",
            "max": 10240
        }, {
            "name": "stderr",
            "max": 10240
        }],
        "cpuLimit": 10000000000,
        "memoryLimit": 104857600,
        "procLimit": 50,
        "copyIn": {
            "Main.class": {
                "fileId": "5UDHKNDA" // 这个缓存文件的 ID 来自上一个请求返回的 fileIds
            }
        }
    }]
}
响应：
[
    {
        "status": "Accepted",
        "exitStatus": 0,
        "time": 53755000,
        "memory": 23064576,
        "runTime": 54205838,
        "files": {
            "stderr": "",
            "stdout": "Hello world\n"
        }
    }
]

**Reply by Gray878** [reply]
报错了："invalid character 'H' after object key:value pair"
请问是什么原因

**Reply by criyle** [reply]
有字符在字符串里面没有转义，检查 `"`

**Reply by ppaet** [reply]
> 报错了："invalid character 'H' after object key:value pair" 请问是什么原因

我的判题service：
https://github.com/ppaet/onlineJudgeExam/blob/eb77995ca112cff9f083536b5f40b2aed0be355e/exam-springboot/src/main/java/com/zeke/service/impl/CodeJudgeServiceImpl.java

**Reply by Gray878** [reply]
感谢回复

**Comment by HC-frank** [comment]
想问一下有没有用过那个windows预编译文件的沙箱的，那个java的环境路径怎么配置？

**Reply by ppaet** [reply]
在docker容器里面安装java，跟外部环境无关，路径看你安装在哪就是哪

**Reply by HC-frank** [reply]
我想尝试弄清那个windows执行文件的路径配置😂

**Reply by criyle** [reply]
可以看一看系统设置里面的环境变量是如何配置的然后复制一份。但是要注意到 windows 沙箱是试验性支持，原理和 chrome 沙箱一致，安全性和在 chrome 里面打开不安全程序一样，生产环境使用请注意风险。
